### DISCUSSION TITLE
OSPF scheme

### DISCUSSION THREAD

**Discussion by junhyeok-yoon** [ANLAB-KAIST/KENSv3#236] [Project Q&A 3. Internet Layer (IP/Routing)] [answered]
I wonder about the code scheme.
I know that the OSPF is executed in router-wise manner, which means that each router runs its on OSPF program.
In that sense, I think one router should use the program seperately.

However, in pwospf test, it seems that routers run with in same code. Is it right?
(Korean)
PWOSPFAssignment.cpp에 OSPF라는 global variable을 두고, 그 global variable을 통해 neighbor 등 여러 정보를 관리하려 합니다.
하지만, cpp에 globla variable로 int entry = 0;을 해놓고, initialize()가 불릴 때 마다, entery++;했는데, 첫번째 test를 돌리면, entry가 2가 됩니다.
제 생각에는, 이 프로그램이 각 라우터에서 따로 돌아야하니까 entry가 1이 되어야할거 같습니다.
이 상황이 의도된 것인가요?

**Comment by Pusnow** [comment, marked as answer]
> However, in pwospf test, it seems that routers run with in same code. Is it right?

Yes, they run same code but with **different data**.
You should define router-specific variable/data inside `PWOSPFAssignment` class.

**Reply by junhyeok-yoon** [reply]
Thanks!

**Reply by junhyeok-yoon** [reply]
1. 각 test가 각 program을 따로 실행하지 않고, 여러 test가 진행되는 동안 같은 program이 계속 돌아가는 것 같은데, 이것도 의도된것인가요?
(TestRoutingOSPFTwoNodes에서 두번 initialize()함수가 실행되는데, 그 다음 RoutingEnvOSPFCustomTwoNodes에서도 initiailize()함수가 두번 실행되서, 이 program에서 실행되는 initialize()는 여기까지 4번입니다.(2번, 2번 따로가 아니라)

2. 만약 의도된 것이라면, 이렇게 변화하는 구조나 link cost에도 적응해서 최종 cost를 구해내는게 목표인가요?

3. Size PWOSPFAssignment::pwospfQuery(const ipv4_t &ipv4)는 그럼 어디서 어디까지의 cost를 구해야하는지 밝혀줘야하지 않나요?
이 query가 어느 router ID에거 온거인지, 그래서 어디까지의 minimum cost를 구해야하는지 모르겠습니다.

**Reply by Pusnow** [reply]
> 1. 각 test가 각 program을 따로 실행하지 않고, 여러 test가 진행되는 동안 같은 program이 계속 돌아가는 것 같은데, 이것도 의도된것인가요?
(TestRoutingOSPFTwoNodes에서 두번 initialize()함수가 실행되는데, 그 다음 RoutingEnvOSPFCustomTwoNodes에서도 initiailize()함수가 두번 실행되서, 이 program에서 실행되는 initialize()는 여기까지 4번입니다.(2번, 2번 따로가 아니라)

Yes, but I'll run all test cases separately on grading.

> 2. 만약 의도된 것이라면, 이렇게 변화하는 구조나 link cost에도 적응해서 최종 cost를 구해내는게 목표인가요?

Yes. If you use member variables of `PWOSPFAssignment` class only, there would be no problem.

> 3. Size PWOSPFAssignment::pwospfQuery(const ipv4_t &ipv4)는 그럼 어디서 어디까지의 cost를 구해야하는지 밝혀줘야하지 않나요?
이 query가 어느 router ID에거 온거인지, 그래서 어디까지의 minimum cost를 구해야하는지 모르겠습니다.

Each router has its own `PWOSPFAssignment` instance, so the source is running router and destination is the `ipv4`.

**Reply by junhyeok-yoon** [reply]
아 이제 뭔 말인지 잘 이해한 것 같습니다. 정말 감사합니다!
